blogposts/2022/introducting-migrator-service.md

```diff
@@ -0,0 +1,149 @@
+
+- **Phase 1**: Create new catalog structures for the index, but do not begin to populate them yet. Transactions that started before the completion of this phase will only see (and therefore only write to) the old catalog structures.
+- **Phase 2**: Wait for all transactions starting before the end of **Phase 1** to finish. Then, perform a full-table scan and write the content of each row to the new index structure. Open the new index structure up to writes.
+- **Phase 3**: Wait for all transactions starting before the end of **Phase 2** to finish. Then, perform another full-table scan and write the content of each row inserted during **Phase 2** to the new index structure. The old index structure can now be closed to writes.
+
+There's a particular detail here that deserves special attention in hindsight: concurrent index creation operations block until **all transactions started before them** have either committed or rolled back. Mixed with long-held advisory locks, we encounter this rather nasty situation.
+
+Observe! In `psql` **Process A**, we acquire an advisory lock:
+
+```sql
+[A] sourcegraph@sourcegraph ❯ SELECT pg_advisory_lock(0, 0);
+ pg_advisory_lock
+------------------
+
+(1 row)
+
+Time: 20.616 ms
+```
+
+In `psql` **Process B**, we attempt to acquire an advisory lock:
+
+```sql
+[B] sourcegraph@sourcegraph ❯ SELECT pg_advisory_lock(0, 0);
+
+```
+
+As this lock is already held by another process, this command blocks indefinitely - or until the user gets bored and aborts their session.
+
+Back in `psql` **Process A**, we attempt to create an index concurrently on _literally any table_:
+
+```sql
+[A] sourcegraph@sourcegraph ❯ CREATE INDEX CONCURRENTLY test ON repo(name, deleted_at);
+ERROR:  40P01: deadlock detected
+DETAIL:  Process 2149 waits for ShareLock on virtual transaction 3/13053932; blocked by process 2471.
+Process 2471 waits for ExclusiveLock on advisory lock [16384,0,0,2]; blocked by process 2149.
+HINT:  See server log for query details.
+LOCATION:  DeadLockReport, deadlock.c:1153
+Time: 1107.901 ms (00:01.108)
+```
+
+After a second or so, the Postgres deadlock detector kicks in, notices two processes vying for each other's locks, and cancels the deadlocked index creation operation. This error is a bit dense, but states that the index creation operation is waiting for **Process B**'s attempt to acquire the advisory lock to finish, but the advisory lock acquisition is blocked on **Process A**'s ownership of that lock. Concurrent index creation failures are also a paritcularly nasty failure mode, as they leave the database with a _partially created_ index that exists in the system catalog but is accessible neither for reads nor writes. To resolve this database state, the invalid index has to be dropped and re-created. This is an easy step to forget, if required to be done manually, and forgetting the step can have very severe performance penalties if not detected.
```

Comment:
```suggestion
After a second or so, the Postgres deadlock detector kicks in, notices two processes vying for each other's locks, and cancels the deadlocked index creation operation. This error is a bit dense, but states that the index creation operation is waiting for **Process B**'s attempt to acquire the advisory lock to finish, but the advisory lock acquisition is blocked on **Process A**'s ownership of that lock. 

Concurrent index creation failures are also a particularly nasty failure mode, because they leave the database with a _partially created_ index that exists in the system catalog but is accessible to reads or writes. To resolve this database state, the invalid index has to be dropped and re-created. This is an easy step to forget you need to do it manually, and forgetting the step can have severe performance penalties if you don't detect it.
```